planner/core/prepare_test.go

```diff
@@ -1193,3 +1196,130 @@ func (s *testPlanSerialSuite) TestIssue23671(c *C) {
 	tk.MustQuery("select @@last_plan_from_cache").Check(testkit.Rows("0"))
 }
+
+func (s *testPlanSerialSuite) TestPartitionTable(c *C) {
+	if israce.RaceEnabled {
+		c.Skip("exhaustive types test, skip race test")
+	}
+
+	// enable plan cache
+	store, dom, err := newStoreWithBootstrap()
+	c.Assert(err, IsNil)
+	tk := testkit.NewTestKit(c, store)
+	orgEnable := core.PreparedPlanCacheEnabled()
+	defer func() {
+		dom.Close()
+		err = store.Close()
+		c.Assert(err, IsNil)
+		core.SetPreparedPlanCache(orgEnable)
+	}()
+	core.SetPreparedPlanCache(true)
+	tk.Se, err = session.CreateSession4TestWithOpt(store, &session.Opt{
+		PreparedPlanCache: kvcache.NewSimpleLRUCache(100, 0.1, math.MaxUint64),
+	})
+	c.Assert(err, IsNil)
+
+	// enable partition table dynamic mode
+	tk.MustExec("create database test_plan_cache")
+	tk.MustExec("use test_plan_cache")
+	tk.MustExec("set @@tidb_partition_prune_mode = 'dynamic'")
+
+	type testcase struct {
+		t1Create string
+		t2Create string
+		rowGener func() string
+		varGener func() string
+		query    string
+	}
+	randDateTime := func() string {
+		return fmt.Sprintf("%v-%v-%v %v:%v:%v",
+			1950+rand.Intn(100), 1+rand.Intn(12), 1+rand.Intn(28), // date
+			rand.Intn(24), rand.Intn(60), rand.Intn(60)) // time
+	}
+	testcases := []testcase{
+		{ // hash partition + int
+			"create table t1(a int, b int) partition by hash(a) partitions 20",
+			"create table t2(a int, b int)",
+			func() string { return fmt.Sprintf("(%v, %v)", rand.Intn(100000000), rand.Intn(100000000)) },
+			func() string { return fmt.Sprintf("%v", rand.Intn(100000000)) },
+			`select * from %v where a > ?`,
+		},
+		{ // range partition + int
+			`create table t1(a int, b int) partition by range(a) (
+	partition p0 values less than (20000000),
+	partition p1 values less than (40000000),
+	partition p2 values less than (60000000),
+	partition p3 values less than (80000000),
+	partition p4 values less than (100000000))`,
+			`create table t2(a int, b int)`,
+			func() string { return fmt.Sprintf("(%v, %v)", rand.Intn(100000000), rand.Intn(100000000)) },
+			func() string { return fmt.Sprintf("%v", rand.Intn(100000000)) },
+			`select * from %v where a > ?`,
+		},
+		{ // range partition + varchar
+			`create table t1(a varchar(10), b varchar(10)) partition by range columns(a) (
+	partition p0 values less than ('200'),
+	partition p1 values less than ('400'),
+	partition p2 values less than ('600'),
+	partition p3 values less than ('800'),
+	partition p4 values less than ('9999'))`,
+			`create table t2(a varchar(10), b varchar(10))`,
+			func() string { return fmt.Sprintf(`("%v", "%v")`, rand.Intn(1000), rand.Intn(1000)) },
+			func() string { return fmt.Sprintf(`"%v"`, rand.Intn(1000)) },
+			`select * from %v where a > ?`,
+		},
+		{ // range partition + datetime
+			`create table t1(a datetime, b datetime) partition by range columns(a) (
+	partition p0 values less than ('1970-1-1 00:00:00'),
```

Comment:
Please use [ISO 8601](https://en.wikipedia.org/wiki/ISO_8601) format with prefix 0 for dates, like 1970-01-01.
Also good to test the 'zero' epoch 🌟

Reply:
OK, I'll change it and add more cases about `timestamp` later.